numpy/core/src/umath/matmul.c.src

```diff
@@ -267,19 +267,71 @@ NPY_NO_EXPORT void
 
 /**end repeat**/
 
+
+NPY_NO_EXPORT void
+OBJECT_matmul_inner_noblas(void *_ip1, npy_intp is1_m, npy_intp is1_n,
+                           void *_ip2, npy_intp is2_n, npy_intp is2_p,
+                           void *_op, npy_intp os_m, npy_intp os_p,
+                           npy_intp dm, npy_intp dn, npy_intp dp)                         
+{
+    npy_intp m, n, p;
+    npy_intp ib1_n, ib2_n, ib2_p, ob_p;
+    char *ip1 = (char *)_ip1, *ip2 = (char *)_ip2, *op = (char *)_op;
+
+    ib1_n = is1_n * dn;
+    ib2_n = is2_n * dn;
+    ib2_p = is2_p * dp;
+    ob_p  = os_p * dp;
+
+    PyObject *tmp1, *tmp2 = NULL;
+
+    for (m = 0; m < dm; m++) {
+        for (p = 0; p < dp; p++) {
+            if ( 0 == dn ) return;
+            // The first pass where n == 0 has to be separated from n>0 because
+            // there might not be a "zero" for objects, and even if there is we don't know it.
+            // This also means the degenerate case is not really well-defined.
```

Comment:
It might make sense to simply return an exception here, although I could also see an argument for returning the integer `0` since `@` is inherently numeric.

Reply:
It's tricky. My impression is that `None` is not too bad, but maybe an error is better.